src/Microsoft.Health.Dicom.Core/Features/Retrieve/RetrieveRenderedService.cs

```diff
@@ -0,0 +1,132 @@
+    private readonly ILogger<RetrieveRenderedService> _logger;
+
+    public RetrieveRenderedService(
+        IInstanceStore instanceStore,
+        IFileStore blobDataStore,
+        IDicomRequestContextAccessor dicomRequestContextAccessor,
+        IOptionsSnapshot<RetrieveConfiguration> retrieveConfiguration,
+        RecyclableMemoryStreamManager recyclableMemoryStreamManager,
+        ILogger<RetrieveRenderedService> logger)
+    {
+        EnsureArg.IsNotNull(instanceStore, nameof(instanceStore));
+        EnsureArg.IsNotNull(blobDataStore, nameof(blobDataStore));
+        EnsureArg.IsNotNull(dicomRequestContextAccessor, nameof(dicomRequestContextAccessor));
+        EnsureArg.IsNotNull(retrieveConfiguration?.Value, nameof(retrieveConfiguration));
+        EnsureArg.IsNotNull(recyclableMemoryStreamManager, nameof(recyclableMemoryStreamManager));
+        EnsureArg.IsNotNull(logger, nameof(logger));
+
+        _instanceStore = instanceStore;
+        _blobDataStore = blobDataStore;
+        _dicomRequestContextAccessor = dicomRequestContextAccessor;
+        _retrieveConfiguration = retrieveConfiguration?.Value;
+        _recyclableMemoryStreamManager = recyclableMemoryStreamManager;
+        _logger = logger;
+    }
+
+    public async Task<RetrieveRenderedResponse> RetrieveRenderedImageAsync(RetrieveRenderedRequest request, CancellationToken cancellationToken)
+    {
+        EnsureArg.IsNotNull(request, nameof(request));
+
+        // To keep track of how long render operation is taking
+        Stopwatch sw = new Stopwatch();
+        sw.Start();
+
+        int partitionKey = _dicomRequestContextAccessor.RequestContext.GetPartitionKey();
+        AcceptHeader returnHeader = GetValidRenderAcceptHeader(request.AcceptHeaders);
+
+        try
+        {
+            // this call throws NotFound when zero instance found
+            InstanceMetadata instance = (await _instanceStore.GetInstancesWithProperties(
+                ResourceType.Instance, partitionKey, request.StudyInstanceUid, request.SeriesInstanceUid, request.SopInstanceUid, cancellationToken)).First();
+
+            FileProperties fileProperties = await RetrieveHelpers.CheckFileSize(_blobDataStore, _retrieveConfiguration.MaxDicomFileSize, instance, cancellationToken);
+            Stream stream = await _blobDataStore.GetFileAsync(instance.VersionedInstanceIdentifier.Version, cancellationToken);
+
+            DicomFile dicomFile = await DicomFile.OpenAsync(stream, FileReadOption.ReadLargeOnDemand);
+            DicomPixelData dicomPixelData = dicomFile.GetPixelDataAndValidateFrames(new[] { request.FrameNumber });
+
+            DicomImage dicomImage = new DicomImage(dicomFile.Dataset);
+            using var img = dicomImage.RenderImage(request.FrameNumber);
+            using var sharpImage = img.AsSharpImage();
+            MemoryStream resultStream = _recyclableMemoryStreamManager.GetStream();
```

Comment:
Can we extract this part and capture a general exception if in case Imagesharp or fo-dicom throws different exception and we can return to the user as DicomImageException